Refactoring workflows and encoding profiles

wsmirnow opened the pull request:
The workflows and encoding profiles in opencast has grown in the past. The idea behind it was to provide an example of each feature of Opencast. As we are adding new features with each release, the grows will never stop. The adopters community on the other hand is using the default workflows in production environments. The root cause IMO. is the complexity and difficult to understand this huge set of workflows and encoding profiles. Here I will start my work and reduce the complexity defining an clean and small set of features we (as opencast community) want to provide. 

The goals of refactoring:

* Reduce the amount of workflows and encoding profiles as much as possible. Remove all workflows and encoding profiles that…
  * aren't used any more
  * are there for historical reasons
  * are optional, very specific or uncommonly used
  * all workflows of disabled features (per default) should be removed, maybe put to the documentation of the feature itself
* Keep the default set of predefined workflows and encoding profiles small
  * Structure and reuse workflow parts
  * Keep it small and simple

To achive the goals we have to make some assumptions:

* Media quality selection for delivery formats could and should be done automaticly
* We have to reduce the amount of predefined publication channels. In most production scenarios the set is static and does not change from episode to another. Here I decided to go with Engage/Search Index. But I leave a comment in the workflows, where you can put your prefered publication channels.
* Keep the editor preview for ever. To use some important editor features we have to keep the editor preview and encode the delivery media from source.

karendolan commented:
Hi @wsmirnow, for the "publish-uploaded-assets" workflow, do you recommend that it be moved into Upload Asset feature documentation. Then, if an adopter wanted to use the asset upload feature to augment an existing publication (i.e. captions file or new thumbnail search image) the user would copy and past the workflow example from the documentation into their etc/workflows directory? 
https://github.com/wsmirnow/opencast/blob/t/workflows-refactoring-rebase-develop-updated/etc/listproviders/event.upload.asset.options.properties#L36

wsmirnow commented:
> Hi @wsmirnow, for the "publish-uploaded-assets" workflow, do you recommend that it be moved into Upload Asset feature documentation. Then, if an adopter wanted to use the asset upload feature to augment an existing publication (i.e. captions file or new thumbnail search image) the user would copy and past the workflow example from the documentation into their etc/workflows directory?
> https://github.com/wsmirnow/opencast/blob/t/workflows-refactoring-rebase-develop-updated/etc/listproviders/event.upload.asset.options.properties#L36

Hi @karendolan, yes it would make sense. I will update the documentation here: https://docs.opencast.org/develop/admin/#configuration/admin-ui/asset-upload/#how-to-enable-preconfigured-asset-options

gregorydlogan commented on etc/encoding/opencast.properties:
```diff
@@ -0,0 +1,210 @@
+  -filter:v select=eq(n\\,1)+lt(t\\,#{time}),fps=1,reverse,scale=160:-2 \
+  #{out.dir}/#{out.name}#{out.suffix}
+
+profile.mp4-preview.http.name = preview video
```
Where are `profile.mp4-preview.http` and `profile.mp4-preview.dual` used?

wsmirnow replied:
`mp4-preview.http` is used for editor preview video generation from single stream, same as `mp4-preview.dual.http`  but for dualstream episodes

wsmirnow replied:
The encoding profiles are mostly moved from separate files to one and wasn't changed by me. Some profile names are not  clear what they are doing or where they are used. Should we generally rename these with an well defined pattern?
Generally I don't like the old pattern. What is `http` in the profile name  for?

gregorydlogan replied:
IIRC, there were `http` and `streaming` profiles, back when the streaming files needed to be different formats.  Very much +1 for refactoring the names to be sane given the current code context.

gregorydlogan commented on etc/encoding/opencast.properties:
```diff
@@ -0,0 +1,210 @@
+  -r 1 -frames:v 1 -filter:v yadif,scale=320:-1 \
+  #{out.dir}/#{out.name}#{out.suffix}
+
+profile.composite.http.name = composite
```
Where are `profile.composite.http` and `profile.concat.work` used?

wsmirnow replied:
Both profiles are there for branding operations. The `composite.http` profile is used for rendering an watermark into the video after cutting but before encoding of delivery version. The `composite.http` concatenate the bumper videos with the actually content video.

gregorydlogan replied:
Can we add comments to that effect to the (renamed) profiles then?

wsmirnow replied:
Yes, this his make sense.

gregorydlogan commented on etc/encoding/opencast.properties:
```diff
@@ -0,0 +1,210 @@
+profile.mux-av.prepared.ffmpeg.command = -i #{in.video.path} -i #{in.audio.path} -shortest \
+  -c copy -map 0:v:0 -map 1:a:0 #{out.dir}/#{out.name}#{out.suffix}
+
+# Used by SelectTracks workflow operation
```
To do what?  I can read the ffmpeg, but someone unfamiliar might not.

wsmirnow replied:
Hmm… this is used by [prepare-av](https://docs.opencast.org/develop/admin/#workflowoperationhandlers/prepareav-woh/) operation. But we wouldn't use it with refactored workflows. This profile should be moved to the documentation of the `prepare-av` operation as well.

gregorydlogan replied:
Fine with that.  Let's make sure there is documentation in the encoding profiles regarding what each one does.

gregorydlogan commented on etc/encoding/opencast.properties:
```diff
@@ -0,0 +1,210 @@
+#   Since this is used by multiple workflow operation handlers, the encoding
+#   profile is duplicated to allow an independent configuration.
+
+# Used by PrepareAV workflow operation
```
To do what?  I can read the ffmpeg, but someone unfamiliar might not.

gregorydlogan commented on etc/encoding/opencast.properties:
```diff
@@ -0,0 +1,210 @@
+profile.mux-av.work.input = stream
+profile.mux-av.work.output = visual
+profile.mux-av.work.suffix = -work.#{in.video.suffix}
+profile.mux-av.work.ffmpeg.command = -i #{in.video.path} -i #{in.audio.path} -shortest \
```
Isn't this the same command as `profile.mux-av.prepared.ffmpeg.command`?

wsmirnow replied:
The functionality is the same. But the first profile is used by [prepare-av](https://docs.opencast.org/develop/admin/#workflowoperationhandlers/prepareav-woh/)  operation, the other by [select-tracks](https://docs.opencast.org/develop/admin/#workflowoperationhandlers/select-streams-woh/).

gregorydlogan replied:
Oi.  As long as that's deliberate I'm fine with it :)

gregorydlogan commented:
Conflicts!

wsmirnow committed:
Refactoring workflows and encoding profiles

Reduce the complexity defining an clean and small set of features we (as opencast community) want to provide.

The goals of refactoring:

* Reduce the amount of workflows and encoding profiles as much as possible. Remove all workflows and encoding profiles that…
  * aren't used any more
  * are there for historical reasons
  * are optional, very specific or uncommonly used
  * all workflows of disabled features (per default) should be removed, maybe put to the documentation of the feature itself
* Keep the default set of predefined workflows and encoding profiles small
  * Structure and reuse workflow parts
  * Keep it small

To achive the goals we have to make some assumptions:

* Media quality selection for delivery formats could and should be done automaticly
* We have to reduce the amount of predefined publication channels. In most production scenarios the set is static and does not change from episode to another. Here I decided to go with Engage/Search Index. But I leave a comment in the workflows, where you can put your prefered publication channels.
* Keep the editor preview for ever. To use some important editor features we have to keep the editor preview and encode the delivery media from source.

wsmirnow commented:
Conflicts solved

wsmirnow committed:
Create feed, search and player preview images from 3rd second

In case you have trimmed the video with the editor, the first two seconds are fade in. In this period, the extracted images are a bit darker as the schould be.

gregorydlogan commented:
@wsmirnow conflicts again

wsmirnow commented:
I would update it next week. Let me change this PR to draft.

lkiesow commented:
Plan for testing:

- Testing new workflows
    - [ ] Upload presentation and publish directly
    - [ ] Upload presenter and publish directly
    - [ ] Upload two streams and publish directly
    - [ ] Upload presentation, send to editor and publish from there
    - [ ] Upload presenter, send to editor and publish from there
    - [ ] Upload two streams, send to editor and publish from there
    - [ ] Upload presentation, send to editor, cut out element and publish
    - [ ] Upload presenter, send to editor, cut out element and publish
    - [ ] Upload two streams, send to editor, cut out element and publish
    - [ ] Retract event and publish it again
    - [ ] Republish metadata
    - [ ] Upload via studio (if the new workflow is used directly, needs more tests otherwise)
- Testing the upgrade path
    - [ ] Upload presentation using old workflow, send to editor and publish from there using the new workflow
    - [ ] Upload presenter using old workflow, send to editor and publish from there using the new workflow
    - [ ] Upload two streams using old workflow, send to editor and publish from there using the new workflow
    - [ ] Upload presentation using old workflow, send to editor, cut out element and publish using the new workflow
    - [ ] Upload presenter using old workflow, send to editor, cut out element and publish using the new workflow
    - [ ] Upload two streams using old workflow, send to editor, cut out element and publish using the new workflow
    - [ ] Retract event published using old workflow and publish it again using the new workflow
    - [ ] Republish metadata of event published using the old workflow

gregorydlogan commented:
Has this made any progress internally?

ebbertd commented:
I ran all the listed tests and only _Retract event published using old workflow and publish it again using the new workflow_ does not work. Retracting from engage and publishing to engage again works. But the event is not retracted from the external api and oai-pmh.

> Plan for testing:
> 
>     * Testing new workflows
>       
>       * [ ]  Upload presentation and publish directly
>       * [ ]  Upload presenter and publish directly
>       * [ ]  Upload two streams and publish directly
>       * [ ]  Upload presentation, send to editor and publish from there
>       * [ ]  Upload presenter, send to editor and publish from there
>       * [ ]  Upload two streams, send to editor and publish from there
>       * [ ]  Upload presentation, send to editor, cut out element and publish
>       * [ ]  Upload presenter, send to editor, cut out element and publish
>       * [ ]  Upload two streams, send to editor, cut out element and publish
>       * [ ]  Retract event and publish it again
>       * [ ]  Republish metadata
>       * [ ]  Upload via studio (if the new workflow is used directly, needs more tests otherwise)
> 
>     * Testing the upgrade path
>       
>       * [ ]  Upload presentation using old workflow, send to editor and publish from there using the new workflow
>       * [ ]  Upload presenter using old workflow, send to editor and publish from there using the new workflow
>       * [ ]  Upload two streams using old workflow, send to editor and publish from there using the new workflow
>       * [ ]  Upload presentation using old workflow, send to editor, cut out element and publish using the new workflow
>       * [ ]  Upload presenter using old workflow, send to editor, cut out element and publish using the new workflow
>       * [ ]  Upload two streams using old workflow, send to editor, cut out element and publish using the new workflow
>       * [ ]  Retract event published using old workflow and publish it again using the new workflow
>       * [ ]  Republish metadata of event published using the old workflow

KatrinIhler commented:
@wsmirnow What are the plans for this one?

mtneug reviewed (changes requested):
Opencast has some encoding profiles hardcoded. I know of the following:

* [`SelectStreamsWorkflowOperationHandler.java`](https://github.com/opencast/opencast/blob/9c1108a465084360076915c82fc93bb600ece16a/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/SelectStreamsWorkflowOperationHandler.java#L62-L68)
* [`PrepareAVWorkflowOperationHandler.java`](https://github.com/opencast/opencast/blob/9c1108a465084360076915c82fc93bb600ece16a/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/PrepareAVWorkflowOperationHandler.java#L61-L70)

In particular these are the profiles required:

```
av.prepared
mux-av.prepared
audio-only.prepared
video-only.prepared
video-only.work
audio-only.work
mux-av.work
```

Some of those are currently missing.

Additionally, there is now a `HLS-streaming-movies.properties` file, you might want to merge with `opencast.properties`.

EDIT

We might want to rename those profiles to

```
av.prepare-av
mux-av.prepare-av
audio-only.prepare-av
video-only.prepare-av

mux-av.select-streams
audio-only.select-streams
video-only.select-streams
```

or something similar.

KatrinIhler commented:
Since this is now worked on in #2573, I'm gonna close this.
